source/extensions/common/async_files/async_file_handle.h

```diff
@@ -0,0 +1,116 @@
+
+  // Enqueues an action to close the currently open file.
+  // It is an error to close on an AsyncFileContext that does not have a file open.
+  // There must not already be an action queued for this handle.
+  virtual void close(std::function<void(absl::Status)> on_complete) = 0;
+
+  // Enqueues an action to read from the currently open file, at position offset, up to the number
+  // of bytes specified by length. The size of the buffer passed to on_complete informs you if less
+  // than the requested amount was read. It is an error to read on an AsyncFileContext that does not
+  // have a file open. There must not already be an action queued for this handle.
+  virtual void read(off_t offset, size_t length,
+                    std::function<void(absl::StatusOr<std::unique_ptr<Envoy::Buffer::Instance>>)>
+                        on_complete) = 0;
+
+  // Enqueues an action to write to the currently open file, at position offset, the bytes contained
+  // by contents. It is an error to call write on an AsyncFileContext that does not have a file
+  // open.
+  //
+  // This call consumes the 'contents' buffer immediately, by move, so it is safe to discard the
+  // buffer after the call, and is not safe to assume it still contains the same data.
+  //
+  // on_complete is called with the number of bytes written on success.
+  // There must not already be an action queued for this handle.
+  virtual void write(Envoy::Buffer::Instance& contents, off_t offset, // NOLINT(runtime/references)
+                     std::function<void(absl::StatusOr<size_t>)> on_complete) = 0;
+
+  // Enqueues a no-op action. This is useful in the case that you have a file action you wish to
+  // perform, but if it can't be performed in a timely fashion you might decide to back out of
+  // sending it.
+  // Using whenReady, you can wait until your request reaches the front of the thread pool queue
+  // before issuing it.
+  // There must not already be an action queued for this handle.
+  virtual void whenReady(std::function<void(absl::Status)> on_complete) = 0;
+
+  // Creates a new AsyncFileHandle referencing the same file.
+  // This is not asynchronous like the other file operations, as it is never a time-consuming
+  // operation. Note that a file handle duplicated in this way shares positioning and permissions
+  // with the original. Since AsyncFileContext functions are all position-explicit, this should not
+  // matter.
+  //
+  // The user should avoid using duplicate while a close, open or create operation is in flight on
+  // the source handle, to avoid provoking a race (the underlying file operation will provide an
+  // appropriate error in the event that a race goes the wrong way).
+  virtual absl::StatusOr<std::shared_ptr<AsyncFileContext>> duplicate() = 0;
+
+protected:
+  virtual ~AsyncFileContext() = default;
+};
+
+using AsyncFileHandle = std::shared_ptr<AsyncFileContext>;
```

Comment:
Usually we just add suffixes to type aliases:
```c++
using AsyncFileContextSharedPtr = std::shared_ptr<AsyncFileContext>;
```
or
```c++
using AsyncFileHandleSharedPtr = std::shared_ptr<AsyncFileHandle>;
```

Reply:
The idea here is that the client only ever uses an AsyncFileHandle, like it should be mostly opaque to the user that it's a shared_ptr, and they shouldn't really have any way to get hold of one that *isn't* a shared_ptr (which is important because the implementation uses std::shared_from_this so it'd cause real problems if the client gets hold of it in any other form).

An earlier implementation did this by having an AsyncFileHandle just be a thin wrapper around a private shared_ptr<AsyncFileContext> so the client could genuinely be completely isolated from the implementation, but the additional layer of repetition made that seem like a bad effort:value ratio.